src/librustc_trans/abi.rs

```diff
@@ -133,32 +134,279 @@ impl ArgAttributes {
     }
 }
 
+pub trait CastTarget {
+    fn llvm_type(&self, ccx: &CrateContext) -> Type;
+}
+
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+pub enum RegKind {
+    Integer,
+    Float,
+    Vector
+}
+
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+pub struct Reg {
+    pub kind: RegKind,
+    pub size: Size,
+}
+
+macro_rules! reg_ctor {
+    ($name:ident, $kind:ident, $bits:expr) => {
+        pub fn $name() -> Reg {
+            Reg {
+                kind: RegKind::$kind,
+                size: Size::from_bits($bits)
+            }
+        }
+    }
+}
+
+impl Reg {
+    reg_ctor!(i8, Integer, 8);
+    reg_ctor!(i16, Integer, 16);
+    reg_ctor!(i32, Integer, 32);
+    reg_ctor!(i64, Integer, 64);
+
+    reg_ctor!(f32, Float, 32);
+    reg_ctor!(f64, Float, 64);
+}
+
+impl CastTarget for Reg {
+    fn llvm_type(&self, ccx: &CrateContext) -> Type {
+        match self.kind {
+            RegKind::Integer => Type::ix(ccx, self.size.bits()),
+            RegKind::Float => {
+                match self.size.bits() {
+                    32 => Type::f32(ccx),
+                    64 => Type::f64(ccx),
+                    _ => bug!("unsupported float: {:?}", self)
+                }
+            }
+            RegKind::Vector => {
+                // Try to pick an integer at least twice as small.
```

Comment:
What's the motivation for this? AFAICT this seems to be used only for SSE/AVX on amd64 SysV, so the `if` statement can never be hit. If LLVM really does not care about the type of the vector elements as long as it has the right length, why not use `[u8; N]` everywhere? Or `[u64; N]` because all SIMD is a multiple of 64 bits.